core/arch/riscv/kernel/thread_optee_abi_rv.S

```diff
@@ -0,0 +1,128 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright 2023 NXP
+ */
+
+#include <asm.S>
+#include <generated/asm-defines.h>
+#include <keep.h>
+#include <kernel/thread.h>
+#include <kernel/thread_private.h>
+#include <riscv.h>
+#include <riscv_macros.S>
+#include <tee/optee_abi.h>
+#include <tee/teeabi_opteed.h>
+#include <tee/teeabi_opteed_macros.h>
+
+
+/*
+ * Implement based on the transport method used to communicate between
+ * untrusted domain and trusted domain. It could be an SBI/ECALL-based to
+ * a security monitor running in M-Mode and panic or messaging-based across
+ * domains where we return to a messaging callback which parses and handles
+ * messages.
+ */
+LOCAL_FUNC thread_return_from_nsec_call , :
+	/* Implement */
+	j	.
+END_FUNC thread_return_from_nsec_call
+
+FUNC thread_std_abi_entry , :
+	jal	__thread_std_abi_entry
+
+	/* Save return value */
+	mv	s0, a0
+
+	/* Disable all interrupts */
+	csrc	CSR_XSTATUS, CSR_XSTATUS_IE
+
+	/* Switch to temporary stack */
+	jal	thread_get_tmp_sp
+	mv	sp, a0
+
+	/*
+	 * We are returning from thread_alloc_and_run()
+	 * set thread state as free
+	 */
+	jal	thread_state_free
+
+	/* Restore __thread_std_abi_entry() return value */
+	mv	a1, s0
+	li	a2, 0
+	li	a3, 0
+	li	a4, 0
+	li	a0, TEEABI_OPTEED_RETURN_CALL_DONE
+
+	/* Return to untrusted domain */
+	jal	thread_return_from_nsec_call
+END_FUNC thread_std_abi_entry
+
+/* void thread_rpc(uint32_t rv[THREAD_RPC_NUM_ARGS]) */
+FUNC thread_rpc , :
```

Comment:
PR https://github.com/OP-TEE/optee_os/pull/6184 fixes SP usage of `thread_rpc()`.
Maybe merge 6184 first and then move it here.

Reply:
A rebase would work also.